types/index.d.ts

```diff
@@ -352,8 +352,34 @@ declare module 'found' {
         }) => React.ReactNode);
   }
 
-  class Link extends React.Component<LinkProps> {
+  type PropsOf<
+    Tag extends React.ReactType
+  > = Tag extends keyof JSX.IntrinsicElements
+    ? JSX.IntrinsicElements[Tag]
+    : Tag extends React.SFC<infer Props>
+    ? Props & React.Attributes
+    : Tag extends React.ComponentClass<infer Props2>
+    ? (Tag extends new (...args: any[]) => infer Instance
+        ? Props2 & React.ClassAttributes<Instance>
+        : never)
+    : never;
+
+  type ReplaceProps<Inner extends React.ReactType, P> = Omit<
+    PropsOf<Inner>,
+    keyof P
+  > &
+    P;
+
+  type LinkProps<As extends React.ReactType> = ReplaceProps<
+    As,
+    BaseLinkProps
+  > & { as?: As };
+
+  class Link<As extends React.ReactType = 'a'> extends React.Component<
+    LinkProps<As>
+  > {
     onClick: (event: React.SyntheticEvent<any>) => void;
+    readonly props: LinkProps<As>;
```

Comment:
is this some magical thing for functional children?

note that the signature of the props passed to children is not the same as what `<Link>` takes; it's actually only

```ts
href: string
active: boolean
onClick: () => ...
```

Reply:
not, this overrides `readonly props: Readonly<P> & Readonly<{ children?: ReactNode }>;` which has the wrong `children` type